src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/Strategy.kt

```diff
@@ -22,18 +25,130 @@ import org.objectweb.asm.ClassVisitor
         throw UnsupportedOperationException("$javaClass strategy does not transform classes")
     }
 
-    abstract fun run(): LincheckFailure?
+    /**
+     * Sets the internal state of strategy to run the next invocation.
+     *
+     * @return true if there is next invocation to run, false if all invocations have been studied.
+     */
+    open fun nextInvocation(): Boolean = true
 
+    /**
+     * Initializes the invocation.
+     * Should be called before each call to [runInvocation].
+     */
+    open fun initializeInvocation() {}
+
+    /**
+     * Runs the current invocation and returns its result.
+     *
+     * Should be called after [initializeInvocation] and only if previous call to [nextInvocation] returned `true`:
+     *
+     * ```kotlin
+     *  with(strategy) {
+     *      if (nextInvocation()) {
+     *          initializeInvocation()
+     *          runInvocation()
+     *      }
+     *  }
+     * ```
+     *
+     * For deterministic strategies, consecutive calls to [runInvocation]
+     * (without intervening [nextInvocation] calls)
+     * should run the same invocation, leading to the same results.
+     *
+     * @return the result of the invocation run.
+     */
+    abstract fun runInvocation(): InvocationResult
+
+    /**
+     * Tries to construct the trace leading to the given invocation result.
+     *
+     * @param result The invocation result.
+     * @return The collected trace, or null if it was not possible to collect the trace.
+     */
+    open fun tryCollectTrace(result: InvocationResult): Trace? = null
+
+    /**
+     * This method is called before the execution of a specific scenario part.
+     *
+     * @param part The execution part that is about to be executed.
+     */
     open fun beforePart(part: ExecutionPart) {}
 
     /**
-     * Is invoked before each actor execution.
+     * This method is called before each actor execution.
+     *
+     * @param iThread the thread index on which the actor is starting
      */
     open fun onActorStart(iThread: Int) {}
+
+    /**
+     * Closes the strategy and releases any resources associated with it.
+     */
+    override fun close() {}
+}
+
+/**
+ * Runs one Lincheck's test iteration with the given strategy and verifier.
+ *
+ * @param iteration the id of the iteration.
+ * @param invocationsBound number of invocations to run.
+ * @param verifier the verifier to be used.
+ *
+ * @return the failure, if detected, null otherwise.
+ */
+fun Strategy.runIteration(iteration: Int, invocationsBound: Int, verifier: Verifier): LincheckFailure? {
+    var spinning = false
+    for (invocation in 0 until invocationsBound) {
+        if (!(spinning || nextInvocation()))
+            return null
+        spinning = false
+        initializeInvocation()
+        val failure = run {
+            val result = runInvocation()
+            spinning = (result is SpinCycleFoundAndReplayRequired)
```

Comment:
This logic is specific to the model checking mode. Can we encapsulate it in the `runInvocation` logic?

Reply:
It depend on whether we want to count invocations with `SpinCycleFoundAndReplayRequired` in the total number of invocations. 
If we want, then, unfortunately, this check has to remain here. 
If not, the replay logic can be moved into model checking mode.

Reply:
Let's do count this case as an invocation, this is just a cost for detecting spin cycles